Stop editing scenario when clicking outside

rajadain opened the pull request:
## Overview

Clicking outside the edit area when renaming scenarios was not canceling the edit operation. Now it does.

## Testing Instructions

 1. Open a scenario
 2. Start renaming a scenario, but don't make any changes to the name
 3. Click anywhere outside. The scenario tab should return to its non-editing state.
 4. Start renaming a scenario and make changes.
 5. Click anywhere outside. Your changes should be saved and the tab should return to its non-editing state.

## Demo

![mmw-rename-scenario](https://cloud.githubusercontent.com/assets/1430060/8918687/a136f8dc-3487-11e5-8139-0318e81c07b7.gif)

Connects #510

lliss commented on src/mmw/js/src/modeling/views.js:
```diff
@@ -342,9 +342,11 @@ var ScenarioTabPanelView = Marionette.ItemView.extend({
                 // Don't add line returns to the text.
                 e.preventDefault();
+                
+                var new_name = $(this).text();
```
prefer camelCase variable names.

lliss commented on src/mmw/js/src/modeling/views.js:
```diff
@@ -352,7 +354,13 @@ var ScenarioTabPanelView = Marionette.ItemView.extend({
 
         this.ui.nameField.on('blur', function() {
-            self.model.collection.updateScenarioName(self.model, $(this).text());
+            var new_name = $(this).text();
```
Lines are nearly identical as above. Might be able to DRY it up.

rajadain commented:
Published new version a8ac7fb incorporating @lliss's suggestions

rajadain commented on src/mmw/js/src/modeling/views.js:
```diff
@@ -326,7 +326,14 @@ var ScenarioTabPanelView = Marionette.ItemView.extend({
-        var self = this;
+        var self = this,
+            setScenarioName = function(name) {
+                if (name && name !== '' && name !== self.model.get('name')) {
```
Empty string is always _falsy_ so the second check is redundant. Fixing ...

rajadain replied:
Also, we [check if the name is the same inside `updateScenarioName`](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/js/src/modeling/models.js#L511) so checking it here in the `if` condition is redundant. Removing that too ...

hectcastro commented:
@azavea-bot rebuild

lewfish commented:
+1, tested.

rajadain commented:
@caseypt since you wrote the original issue, could you verify that this fix is what you were looking for?

caseypt commented:
@rajadain Sure! Will take a look soon.

caseypt commented on src/mmw/js/src/modeling/views.js:
```diff
@@ -326,7 +326,14 @@ var ScenarioTabPanelView = Marionette.ItemView.extend({
     renameScenario: function() {
-        var self = this;
+        var self = this,
+            setScenarioName = function(name) {
```
Does this need to be defined within `renameScenario`? If not, consider pulling it out.

rajadain replied:
It doesn't _need_ to be, but since we're packaging local functionality for use twice locally (in response to https://github.com/WikiWatershed/model-my-watershed/pull/619#discussion_r35756091), I kept it local. I suppose it could be useful to have this outside, but I didn't have an immediate use case for that and wanted to contain my change to have a small footprint (keep it within `renameScenario`). If this is not satisfactory, then I can move it out.

caseypt replied:
Okay cool. That's fine with me

caseypt commented:
If you enter a scenario name that already exists, like "Current Conditions", and then click away. It leaves edit mode and shows the name as "Current Conditions". But then if you click on one of the other scenario tabs, it will rerender the tabs and show that the scenario name wasn't changed (which is correct). If the validation fails, we should reset the value in the tab when clicking away.

I thought a prompt showed up if you used an invalid name of the scenario, but it looks like that isn't happening on staging either.

rajadain commented:
Hmm good catch. This is likely a problem with the name checking logic inside `updateScenarioName` (as evidenced by this also being on Staging) but I will investigate and fix as part of this PR.

rajadain commented:
Could use one final look through. I'm now reusing the logic available in `updateScenarioName` by making it always return a value. If the return value is falsy we manually render. I also added an alert box asking the user for unique names. Previously we only logged to the console.

We should probably make a card for adding a custom alert box.

lliss commented on src/mmw/js/src/modeling/models.js:
```diff
@@ -517,10 +517,13 @@ var ScenariosCollection = Backbone.Collection.extend({
         });
 
         if (match) {
+            window.alert("There is another scenario with the same name. " +
```
Can we use one of the modals for alerts?

lliss replied:
Strike that just saw the note in your comment.

lliss commented:
@azavea-bot rebuild

lliss commented on src/mmw/js/src/modeling/models.js:
```diff
@@ -517,10 +517,13 @@ var ScenariosCollection = Backbone.Collection.extend({
+            return false;
         } else if (model.get('name') !== newName) {
-            model.set('name', newName);
+            return model.set('name', newName);
```
If I don't need the returned value on the other side, I personally try to return boolean rather than truthy and falsey values.

rajadain replied:
This should hopefully take care of the unlikely edge case when a `model.set` fails (for example, if there are [validation errors](http://backbonejs.org/docs/backbone.html#section-68))

lliss commented:
+1

caseypt commented:
+1

rajadain commented:
@azavea-bot rebuild

rajadain commented:
@azavea-bot rebuild

rajadain commented:
The tests were stalling because of the `window.alert` that was added recently. Rebased on top of `develop`, and updated tests in 5b1c198 to handle that.

@lewfish could you please have a look at the tests and see if they are stylistically consistent?

rajadain commented:
Fixed lint and repushed d0db809

caseypt commented:
Cool fix.

rajadain commented:
Added #645 for adding an alert modal.

lewfish commented:
Taking a look.

lewfish commented:
+1, tested.
